sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,198 @@
+"""Perform actions on Active Directory."""
+
+from __future__ import annotations
+
+from pytest_mh import MultihostHost, MultihostUtility
+from pytest_mh.conn import ProcessResult
+
+__all__ = [
+    "AdcliUtils",
+]
+
+
+class AdcliUtils(MultihostUtility[MultihostHost]):
+    """
+    Interface to adcli utility.
+
+    .. code-block:: python
+        :caption: Example usage
+
+        @pytest.mark.topology(KnownTopologyGroup.AnyAD)
+        def test_adcli_join(client: Client, provider: GenericADProvider):
+            cred = provider.host.adminpw
+            r = client.adcli.join(provider.host.domain, ["--domain-controller", provider.host.hostname], password=cred)
+            assert provider.host.domain in r.stdout, "adcli failed to join the client!"
+
+    .. note::
+
+       This utility will not revert any changes. It relies on AD host topology for clean up.
+       For methods requiring an authentication, --stdin-password(-W) is a default. Setting krb=True will enable
+       kerberos based authentication.
+    """
+
+    def info(self, domain: str, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        Discover AD domain.
+
+        :param domain: domain.
+        :type domain: str,
+        :param args: additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
```

Comment:
Need to add return and rtype lines to the docstrings also.

Reply:
done